File: backend/src/Models/ServiceProvider.php
<?php

namespace Kivou\Models;

class ServiceProvider
{
    public int $id;
    public ?int $ownerUserId = null;
    public string $name;
    public string $email;
    public string $phone;
    public ?string $photoUrl;
    public string $description;
    public array $categories = [];
    public float $rating = 0.0;
    public int $reviewsCount = 0;
    public float $pricePerHour = 0.0;
    public float $latitude = 0.0;
    public float $longitude = 0.0;
    public array $gallery = [];
    public array $availableDays = [];
    public ?string $workingStart = null;
    public ?string $workingEnd = null;
    public bool $isAvailable = true;

    public static function fromRow(array $r): self
    {
        $p = new self();
        $p->id = (int)$r['id'];
        if (isset($r['owner_user_id'])) {
            $p->ownerUserId = $r['owner_user_id'] !== null ? (int)$r['owner_user_id'] : null;
        }
        $p->name = $r['name'];
        $p->email = $r['email'];
        $p->phone = $r['phone'];
        $p->photoUrl = $r['photo_url'] ?? null;
        $p->description = $r['description'] ?? '';
        $p->categories = self::toArray($r['categories'] ?? '');
        $p->rating = (float)$r['rating'];
        $p->reviewsCount = (int)($r['reviews_count'] ?? 0);
        $p->pricePerHour = (float)$r['price_per_hour'];
        $p->latitude = (float)$r['latitude'];
        $p->longitude = (float)$r['longitude'];
        $p->gallery = self::toArray($r['gallery'] ?? '');
        $p->availableDays = self::toArray($r['available_days'] ?? '');
        $p->workingStart = $r['working_start'] ?? null;
        $p->workingEnd = $r['working_end'] ?? null;
        $p->isAvailable = (int)($r['is_available'] ?? 1) === 1;
        return $p;
    }

    public function json(): array
    {
        return [
            'id' => $this->id,
            'owner_user_id' => $this->ownerUserId,
            'name' => $this->name,
            'email' => $this->email,
            'phone' => $this->phone,
            'photo_url' => $this->photoUrl,
            'description' => $this->description,
            'categories' => $this->categories,
            'rating' => $this->rating,
            'reviews_count' => $this->reviewsCount,
            'price_per_hour' => $this->pricePerHour,
            'latitude' => $this->latitude,
            'longitude' => $this->longitude,
            'gallery' => $this->gallery,
            'available_days' => $this->availableDays,
            'working_start' => $this->workingStart,
            'working_end' => $this->workingEnd,
            'is_available' => $this->isAvailable ? 1 : 0,
        ];
    }

    private static function toArray(string $s): array
    {
        return array_values(array_filter(array_map('trim', explode(',', $s))));
    }
}

File: lib/models/service_provider.dart
/// Modèle représentant un prestataire de service
class ServiceProvider {
  final String id;
  final String name;
  final String email;
  final String phone;
  final String photoUrl;
  final String description;
  final List<String> categories;
  final double rating;
  final int reviewsCount;
  final double pricePerHour;
  final double latitude;
  final double longitude;
  final List<String> gallery;
  final List<String> availableDays; // ['monday', 'tuesday', ...]
  final TimeSlot workingHours;
  final bool isAvailable;
  final DateTime createdAt;

  ServiceProvider({
    required this.id,
    required this.name,
    required this.email,
    required this.phone,
    required this.photoUrl,
    required this.description,
    required this.categories,
    required this.rating,
    required this.reviewsCount,
    required this.pricePerHour,
    required this.latitude,
    required this.longitude,
    required this.gallery,
    required this.availableDays,
    required this.workingHours,
    required this.isAvailable,
    required this.createdAt,
  });

  /// Distance depuis un point donné (en kilomètres)
  double distanceFrom(double lat, double lng) {
    // Calcul simplifié de distance (pour production, utiliser geolocator)
    const double earthRadius = 6371; // Rayon de la Terre en km
    double dLat = _toRadians(latitude - lat);
    double dLng = _toRadians(longitude - lng);

    double a =
        (dLat / 2) * (dLat / 2) + (dLng / 2) * (dLng / 2) * (lat * (latitude));
    double c = 2 * (a);

    return earthRadius * c;
  }

  double _toRadians(double degree) {
    return degree * (3.14159 / 180);
  }

  ServiceProvider copyWith({
    String? name,
    String? email,
    String? phone,
    String? photoUrl,
    String? description,
    List<String>? categories,
    double? rating,
    int? reviewsCount,
    double? pricePerHour,
    double? latitude,
    double? longitude,
    List<String>? gallery,
    List<String>? availableDays,
    TimeSlot? workingHours,
    bool? isAvailable,
  }) {
    return ServiceProvider(
      id: id,
      name: name ?? this.name,
      email: email ?? this.email,
      phone: phone ?? this.phone,
      photoUrl: photoUrl ?? this.photoUrl,
      description: description ?? this.description,
      categories: categories ?? this.categories,
      rating: rating ?? this.rating,
      reviewsCount: reviewsCount ?? this.reviewsCount,
      pricePerHour: pricePerHour ?? this.pricePerHour,
      latitude: latitude ?? this.latitude,
      longitude: longitude ?? this.longitude,
      gallery: gallery ?? this.gallery,
      availableDays: availableDays ?? this.availableDays,
      workingHours: workingHours ?? this.workingHours,
      isAvailable: isAvailable ?? this.isAvailable,
      createdAt: createdAt,
    );
  }
}

/// Créneau horaire
class TimeSlot {
  final String start; // Format "HH:mm"
  final String end; // Format "HH:mm"

  TimeSlot({
    required this.start,
    required this.end,
  });

  @override
  String toString() => '$start - $end';
}
